p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,485 @@ where
+        }
+
+        block_ids.truncate(requested_blocks_limit - peer_state.num_blocks_to_send);
+        peer_state.num_blocks_to_send += block_ids.len();
+        debug_assert!(peer_state.num_blocks_to_send <= requested_blocks_limit);
+        self.blocks_queue.extend(block_ids.into_iter().map(|id| (peer, id)));
+
+        Ok(())
     }
 
-    /// Validate incoming header response
-    async fn validate_header_response(
+    async fn handle_header_response(
         &mut self,
-        peer_id: &T::PeerId,
+        peer: T::PeerId,
         headers: Vec<BlockHeader>,
-    ) -> crate::Result<Option<BlockHeader>> {
-        ensure!(
-            headers.len() <= HEADER_LIMIT,
-            P2pError::ProtocolError(ProtocolError::InvalidMessage),
-        );
+    ) -> Result<()> {
+        log::debug!("Headers response from peer {peer}");
 
-        let peer = self
+        let peer_state = self
             .peers
-            .get_mut(peer_id)
+            .get_mut(&peer)
             .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?;
+        if !peer_state.known_headers.is_empty() {
+            return Err(P2pError::ProtocolError(ProtocolError::UnexpectedMessage(
+                "headers response",
+            )));
+        }
 
-        // empty response means that local and remote are in sync
-        if headers.is_empty() {
-            return Ok(None);
+        if headers.len() > self.p2p_config.msg_header_count_limit.clone().into() {
+            return Err(P2pError::ProtocolError(
+                ProtocolError::HeadersLimitExceeded(
+                    headers.len(),
+                    self.p2p_config.msg_header_count_limit.clone().into(),
+                ),
+            ));
         }
+        log::trace!("Received headers: {headers:#?}");
 
-        // verify that the first headers attaches to local and chain
-        // and that the received headers are in order
-        match peer.state() {
-            peer::PeerSyncState::UploadingHeaders(ref locator) => {
-                let genesis_id = self.chain_config.genesis_block_id();
-                let mut locator = locator.iter().chain(std::iter::once(&genesis_id));
-                let anchor_point = headers[0].prev_block_id();
-                ensure!(
-                    locator.any(|id| anchor_point == id),
-                    P2pError::ProtocolError(ProtocolError::InvalidMessage),
-                );
-            }
-            _ => return Err(P2pError::ProtocolError(ProtocolError::InvalidMessage)),
+        // We are in sync with this peer.
+        if headers.is_empty() {
+            return Ok(());
         }
 
-        for (a, b) in headers.iter().zip(&headers[1..]) {
-            ensure!(
-                b.prev_block_id() == &a.get_id(),
-                P2pError::ProtocolError(ProtocolError::InvalidMessage),
-            );
+        // Each header must be connected to the previous one.
+        if !headers
+            .iter()
+            .zip(&headers[1..])
```

Comment:
Isn't easier to use the `windows(2)` method in `Vec`?

Reply:
I don't like `std::slice::windows` because this method returns a slice, so I have rewritten this code using `tuple_windows` from `itertools`.